### PR TITLE
Issue/4044 Basic detekt integration

### PR DESCRIPTION
**Pull request by kidinov**
Closes Issue #4044

Following [the post] here is the first step of Detekt integration in our project

paqN3M-jK-p2

It contains:
* Baseline (without formatting)
* Config file
* Detekt plugin connection
* Detekt-formatter plugin connection

Please go through my comments in the  PR

### How to test
Run `./gradlew detektAll` and notice some code formatting violations

@woocommerce/platform-9 someone please take a look as well, just in case

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -3,6 +3,7 @@ buildscript {
     ext.navigationVersion = '2.3.5'
     ext.hiltVersion = '2.35.1'
     ext.wordPressLoginVersion = '58-bd338e6eadb0eadf640f2320c791ca58f86c0cda'
+    ext.detektVersion = '1.15.0'
```
Comment (kidinov) [inline build.gradle:6]: We have to use old Detekt here because of the old version of AGP (4.0.2) we use

https://detekt.github.io/detekt/compatibility.html

@wzieba told me that we can not just update AGP because we use composite builds. @oguzkocer I know that you have been working with this lately. Any plans to make it possible?

File: build.gradle [hunk 2]
```diff
@@ -67,6 +70,16 @@ subprojects {
         classpath = configurations.ktlint
         args "src/**/*.kt", "--reporter=checkstyle,output=${buildDir}/ktlint.xml"
     }
+
+    detekt {
+        config = files("${project.rootDir}/config/detekt/detekt.yml")
+        baseline = file("${project.rootDir}/config/detekt/baseline.xml")
+        buildUponDefaultConfig = true
```
Comment (kidinov) [inline build.gradle]: From the doc:

```
Applies the config files on top of detekt's default config file. `false` by default.
```

It sounds reasonable to me. We stick to the default params and having very short config file which contains only our deviations from the default. Easy to keep up to date. Easier to understand for the newcomers. Seems to be win-win to me. Wdyt?

**Comment by kidinov** [comment]
We currently have 45 formatting issues in the codebase.

```
        MultiLineIfElse - [] at /Users/andrey/StudioProjects/woocommerce-android/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreStatsView.kt:441:17
        SpacingAroundColon - [] at /Users/andrey/StudioProjects/woocommerce-android/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreDateRangeView.kt:17:96
        MultiLineIfElse - [] at /Users/andrey/StudioProjects/woocommerce-android/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/DashboardStatsBarChart.kt:36:13
        MultiLineIfElse - [] at /Users/andrey/StudioProjects/woocommerce-android/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/DashboardStatsBarChart.kt:55:17
        MultiLineIfElse - [] at /Users/andrey/StudioProjects/woocommerce-android/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/RefundSummaryFragment.kt:95:21
        MultiLineIfElse - [] at /Users/andrey/StudioProjects/woocommerce-android/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/RefundSummaryFragment.kt:97:21
        MultiLineIfElse - [] at /Users/andrey/StudioProjects/woocommerce-android/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt:270:21
        MultiLineIfElse - [] at /Users/andrey/StudioProjects/woocommerce-android/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/refunds/IssueRefundViewModel.kt:272:21
etc
```

 From the [detekt doc](https://detekt.github.io/detekt/suppressing-rules.html#formatting-rules-suppression):

```
Please note that rules inside the formatting ruleset can only be suppressed at the file level.
```

Underneath it runs ktlint. I'm not 100% sure why our current ktlint doesn't catch those 45 issues, my guess because we use 3 years old ktlint version so maybe just new checks were added.

Nevertheless, what I propose is to fix those issues in a separate PR (they all minor formatting issues), instead of suppressing them in the code. Agree?

File: config/detekt/baseline.xml [hunk 3]
```diff
@@ -0,0 +1,419 @@
+<?xml version="1.0" ?>
```
Comment (kidinov) [inline config/detekt/baseline.xml:1]: It's an automatically generated file, so no need to touch it with hands, but it represents the violations we currently have in the code. Btw, it's not that many of them.

**Comment by AliSoftware** [comment]
@ParaskP7 I'm wondering if you synchronized here between the work you plan to do on WPAndroid during HACK week and the work done by @kidinov on WCAndroid there? Would be nice to update both Android projects to Detekt + clear baseline, but not sure that the two projects (WC and WP) are in the same state of progress here so might also not make sense to try to mutualize 🤔  In any case I figured it might be worth a ping to let you know and see if we could use the HACK week to align WC and WP on that point 😜

**Comment by ParaskP7** [comment]
👋 @AliSoftware !

Thanks for the ping, much appreciated! 🙏 

> @ParaskP7 I'm wondering if you synchronized here between the work you plan to do on WPAndroid during HACK week and the work done by @kidinov on WCAndroid there?

The work I plan to do on `Detekt` for `WPAndroid` for this hack week, if I will be left with anytime after my work on Lint, would be to clean the baseline by fixing/suppressing those baseline issues, nothing structural, just some extra manual work to clean-up, as much as possible, this baseline of ours. As such, I don't think any synchronisation needed between `WPAndroid` and `WCAndroid` based on my work to be done.

> Would be nice to update both Android projects to Detekt + clear baseline, but not sure that the two projects (WC and WP) are in the same state of progress here so might also not make sense to try to mutualize 🤔

You are right, at this point, and based on the work I will be (potentially) doing on `Detekt` for `WPAndroid` for this hack week, there isn't much to mutualize, apart from the fact that it would be nice if both repos followed the same approach during configuration. As such, @kidinov , please let me know if you in case you follow another path with the configuration.

> In any case I figured it might be worth a ping to let you know and see if we could use the HACK week to align WC and WP on that point 😜

Yes, definitely, it would be totally beneficial for us to align `Detekt` for both, `WPAndroid` and `WCAndroid`, configuration wise, as much as possible. Again, many thanks for the ping @AliSoftware ! 🙏

File: build.gradle (hunk 2 above)
Comment (ParaskP7) [reply build.gradle]: 👋 @kidinov !

On `WPAndroid` I started with having this configuration as per the default, set to `false`. I actually never used `Detekt` so far with that configuration set to `true`, thus I am not sure, at least to a great extend, what this might look like going forward. The reason I had that set to `false` is the fact that with every update, new rules are added to the default configuration, and those rules with be then forced upon us. Instead, I would wanted us to make the effort and update the configuration rules file ourselves, keeping it always up-to-date with the default configuration, with every update, and this way to have more control over all that.

Does this makes sense? am I misunderstanding it or can't explain my reasoning well?

File: build.gradle (hunk 1 above)
Comment (oguzkocer) [reply build.gradle:6]: @kidinov I am actively working on the Gradle upgrade which as @wzieba mentioned needs to happen for multiple libraries/clients at once. We'll want to at least do it for WPAndroid, WCAndroid, Gutenberg and FluxC so we cover the most active codebases. We also have several issues in WPAndroid that needs to be fixed before we can update to Gradle 7.0, so it'll take at least another few days.

Comment (kidinov) [reply build.gradle:6]: @oguzkocer Perfect! 🙏

File: build.gradle (hunk 2 above)
Comment (kidinov) [reply build.gradle]: @ParaskP7 Your explanation is clear, thanks!

But to be honest I see this from a different angle. So let's say someone updates detekt with the new version and this version introduces several cool new rules. With the approach, you are taking there is quite a chance that those new rules will be turned off forever and we won't even notice that we are missing them. With my approach, the CI will just fail (if we don't follow the rules) and a person who does the update must acknowledge the rules and start a discussion if we need to keep them on or off

So it's kind of people make mistakes, machines not, so why not to use machines to make our lives easier =)

Comment (ParaskP7) [reply build.gradle]: 👋 @kidinov !

Yes, this totally makes sense, but with this approach the rule file, `detekt.yml` will progressively get out-of-date and it will not be easy for an engineer to notice the diff between the default config and our configuration. I personally like to be explicit with the set of rules that are enabled/disabled as this helps me when I am configuring them on demand (after the update).

Also, imagine this scenario where even after an update, a specific engineer finds themselves with a warning that needs to be fixed. This warning might be some new warning that is not listed in our configuration, but only in the new default, and it can be more complicated than just a `true` or `false` rule, something like the `StringLiteralDuplication`, which has a threshold of `3` and other configurations within. In order for the engineer to be able to reason about that, or even suggest a change (from a threshold of `3` to `4`, or anything else), it will make it more difficult for them, don't you think?

PS: This is just my 2 cents, please don't get discouraged by my reasoning and way of `Detekting`. I have been using that for few years now and this works for me, but I haven't tried this approach of yours, which might have other benefits and can potentially help with easing the update process. We can try to use this approach of yours on `Woo` and my approach on `WordPress`, then we can catch-up and exchange our outcomes, pain-points, etc.

Comment (kidinov) [reply build.gradle]: > but with this approach the rule file, detekt.yml will progressively get out-of-date and it will not be easy for an engineer to notice the diff between the default config and our configuration. I personally like to be explicit with the set of rules that are enabled/disabled as this helps me when I am configuring them on demand (after the update).

Hmm, probably we are not on the same page here, because from my understanding with my approach `detekt.xml` will be in sync all the time with default configuration because `detekt.xml` contains the ONLY diff with default config. So ideally it will be always empty (which means we use default config) unless we as a team will decide that we should relax or rune some rules if they don't work for us or hard to follow. 

And the opposite, In the approach when you are trying to keep all the possible rules which detekt supports in your config, I think it's almost impossible to understand what rules you put there has default configuration and what were changed by someone.

> Also, imagine this scenario where even after an update, a specific engineer finds themselves with a warning that needs to be fixed. This warning might be some new warning that is not listed in our configuration, but only in the new default, and it can be more complicated than just a true or false rule, something like the StringLiteralDuplication, which has a threshold of 3 and other configurations within. In order for the engineer to be able to reason about that, or even suggest a change (from a threshold of 3 to 4, or anything else), it will make it more difficult for them, don't you think?

This will happen only in cases when the engineer explicitly updates Detekt version and in my perception, this engineer should try to fix the violations firstly. The defaults can be changed only after a discussion with a team. If we are talking about the specific situation when detekt gives a false positive then @Supress annotation should be used I guess.

In general, this is a similar situation as we do with ktlint at the moment and we have there only:

```
[*.{kt,kts}]
max_line_length=120
indent_style=space
indent_size=4
```

So basically we list only our specific rules which are different with defaults. Not all the rules. The same goes with checkstyle rules. 

Also, it looks like that detekt developer also [thinks that it should be true by default](https://github.com/detekt/detekt-compiler-plugin/issues/13)

> PS: This is just my 2 cents, please don't get discouraged by my reasoning and way of Detekting. I have been using that for few years now and this works for me, but I haven't tried this approach of yours, which might have other benefits and can potentially help with easing the update process. We can try to use this approach of yours on Woo and my approach on WordPress, then we can catch-up and exchange our outcomes, pain-points, etc.

Sure thing! Always nice to have a different angle on an issue

Comment (ParaskP7) [reply build.gradle]: 👋 @kidinov !

> Hmm, probably we are not on the same page here, because from my understanding with my approach detekt.xml will be in sync all the time with default configuration because detekt.xml contains the ONLY diff with default config.

True. 🤔 

If this is how it works, the `detekt.yml` to **ONLY** contain the diff, then I actually think that this is a **GOOD** idea. For some reason I thought it doesn't work like that, please excuse my ignorance as I never tried this configuration, maybe way back, but can't remember now.

Let's try your approach with `Woo` and this other (current) approach on `WordPress` and gather some insights. Later on and if you are happy with this approach, I can make the switch on `WordPress` as well. 🙏 

> Also, it looks like that detekt developer also thinks that it should be true by default

Nice finding! 🌟  

This is very interesting, I wonder why this is still `false` by default then, instead of `true`. Is the [specific documentation](https://detekt.github.io/detekt/gradle.html) out-dated? Can you check code-wise if this change has landed with new versions of `Detekt`?

Comment (kidinov) [reply build.gradle]: > This is very interesting, I wonder why this is still false by default then, instead of true. Is the specific documentation out-dated? Can you check code-wise if this change has landed with new versions of Detekt?

I think it's out of date. On [github](https://github.com/detekt/detekt) it's actually true

```
detekt {
    buildUponDefaultConfig = true // preconfigure defaults
```

Comment (ParaskP7) [reply build.gradle]: > I think it's out of date. On github it's actually true

Are you sure this is the default, I think we need to check the code after all? 🤔 

It seems to be that on `GitHub` they are changing it to `true` by explicitly specifying it, otherwise they wouldn't have set it explicitly to `true` and could have depended on the underneath `default` value, just like it is done for the rest. I am not sure if this `preconfigure defaults` comment is meant to say that this is the default `true` configuration or that with this line they are preconfiguring the default `rules`. Tricky right! 😅 

In anyway, I think we have a plan, thanks for grinding this with me, I now have a better understanding of this configuration. 🙏

File: config/detekt/baseline.xml [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+    <ID>TooGenericExceptionCaught:FCMRegistrationIntentService.kt$FCMRegistrationIntentService$e: Exception</ID>
+    <ID>TooGenericExceptionCaught:FileUtils.kt$FileUtils$e: Exception</ID>
+    <ID>TooGenericExceptionCaught:OrderShipmentTrackingHelper.kt$OrderShipmentTrackingHelper$e: Exception</ID>
+    <ID>TooGenericExceptionCaught:PhoneUtils.kt$PhoneUtils$e: Exception</ID>
+    <ID>TooGenericExceptionCaught:ProductDetailViewModel.kt$ProductDetailViewModel$e: Exception</ID>
+    <ID>TooGenericExceptionCaught:StringDateFormatExt.kt$e: Exception</ID>
+    <ID>TooGenericExceptionCaught:TokenProvider.kt$TokenProvider$e: RuntimeException</ID>
+    <ID>TooGenericExceptionCaught:WooLogViewerActivity.kt$WooLogViewerActivity$e: Exception</ID>
+    <ID>TooGenericExceptionThrown:ActionableEmptyView.kt$ActionableEmptyView$throw RuntimeException("$context: ActionableEmptyView must have a title (aevTitle)")</ID>
+    <ID>TooManyFunctions:AddAttributeFragment.kt$AddAttributeFragment : BaseProductFragment</ID>
```
Comment (hichamboushaba) [inline config/detekt/baseline.xml:303]: I think this `TooManyFunctions` rule will get in our way to much, especially in Activities and Fragments where we need a big amount of functions by default.
I believe we should modify the [rule](https://detekt.github.io/detekt/complexity.html#toomanyfunctions) by enabling `ignoreOverridden`, which would ignore all the overridden function from the rule, and then re-calculate the baseline, WDYT?
cc @woocommerce/android-developers for your opinion as well

**Review by hichamboushaba — COMMENTED** [review]
Thanks so much @kidinov for working on this 🥇 
I left one comment regarding the `TooManyFunctions` rule, let's discuss it

File: config/detekt/baseline.xml (hunk 4 above)
Comment (kidinov) [reply config/detekt/baseline.xml:303]: My plan was that a discussion like this will go to phase 2. I will create a P2 where we could comment on the cases like that based on our real day-to-day work. After some time (a few weeks) we will have a general overview of possible modifications we might need. Then based on the collective decisions we would ignore/modify the rules via the detekt.xml file. Should it work, wdyt?

Comment (hichamboushaba) [reply config/detekt/baseline.xml:303]: I think this would work for most of the other rules, but for the rule `TooManFunctions` rule, once we enable the integration, and before getting a chance to discuss and agree on the rules, we may be blocked for any new fragment/activity that we add in the app, and we may need to suppress the error explicitly.
This is just my opinion 🙂, and it's not a big deal, if you think leaving it to later is better, then go for it 👍

Comment (kidinov) [reply config/detekt/baseline.xml:303]: If this will be really striking us in a foot super often then we can have a quick discussion in woo-async and then change configuration right in the PR which is blocked by this. I just don't want to change rules before we definitely know that we can not follow them by one ore more reasons

**Review by hichamboushaba — APPROVED** [review]
LGTM, some defaults have to be discussed by the team later to see if we should stick to them, or modify the rules.
